### PR TITLE
Keyboard navigation

### PR DESCRIPTION
**Pull request by rlisahuang**

### CODE REVIEW

**Comment by rlisahuang** [comment]
Interaction flow:

- Tab to `focus` on the editor
- Press `Enter` to start editing (and you will have an extra `\n` at the beginning`)
- When done, hit `Escape` to continue tab/ shift+tab to other elements
- Next time when your focus is back on the editor again, you can directly start editing; you no longer need to press `Enter`.

**Comment by rlisahuang** [comment]
On the other hand, we sacrifice part of the functionality of `undo`: you can still reset/undo the reset, but you will be able to edit the code during the three second interval, and the change will be recorded (for you to further reset/undo).

The way React handle states is a mess...

**Comment by rlisahuang** [comment]
If you are curious about why you need an unintuitive `Enter` after first tabbing on the editor to start editing, that is due to an internal bug with `use-editable`:
- https://github.com/FormidableLabs/use-editable/issues/16
- https://github.com/FormidableLabs/use-editable/issues/20

**Comment by rlisahuang** [comment]
With this PR #47 can be closed.

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 1]
```diff
@@ -121,19 +123,38 @@ function CodeEditor(props: {
...
     }
 
     const onClickUndo = () => {
-        console.log(tmpCode)
         setCode(tmpCode);
     }
 
-    //   prismIncludeLanguages(Prism);
+
+    document.addEventListener('keydown',
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:140]: you are adding this listener on every render, instead use useEffect to add it and remove it.

Comment (pelikhan) [reply website/src/components/TutorialComponents/CodeBlock.tsx:140]: add event exclusively on the editor instance better?

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 2]
```diff
@@ -165,20 +186,28 @@ function CodeEditor(props: {
...
+                                console.log(e.target)
+                                const selectObj = window.getSelection();
+                                if (selectObj.rangeCount === 0) {
+                                    const range = new Range();
+                                    range.collapse(true);
+                                    selectObj.addRange(range);
+                                }
+                                setHasFocus(true);
+                            }}
+                            onBlur={() => setHasFocus(false)}
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:210]: eslint should have flagged this

**Review by pelikhan — CHANGES_REQUESTED** [review]

File: website/src/components/TutorialComponents/CodeBlock.tsx (hunk 1 above)
Comment (rlisahuang) [reply website/src/components/TutorialComponents/CodeBlock.tsx:140]: I tried `onkeydownCapture` on the editor instance itself but that has some race condition with the `onKeydown` handler of `use-editable` that we do not have control. Capture `keydown` from the very top level would solve the problem.

Comment (rlisahuang) [reply website/src/components/TutorialComponents/CodeBlock.tsx:140]: Not so familiar with the `useEffect` approach, let me see if it produces the same behavior.

File: website/src/components/TutorialComponents/CodeBlock.tsx (hunk 2 above)
Comment (rlisahuang) [reply website/src/components/TutorialComponents/CodeBlock.tsx:210]: the missing `event` parameter?

**Comment by rlisahuang** [comment]
New interaction flow with the requested changes:

- Tab to focus on the editor
- Press Enter to start editing
  - You will see an extra `\n` at the beginning
  - However, once you start typing, the extra `\n` is gone
- When done, hit Escape to continue tab/ shift+tab to other elements
- Next time when your focus is back on the editor again, you can directly start editing; you no longer need to press Enter.

**Comment by rlisahuang** [comment]
Cursor position TODO in #113

File: website/src/components/TutorialComponents/CodeBlock.tsx (hunk 1 above)
Comment (pelikhan) [reply website/src/components/TutorialComponents/CodeBlock.tsx]: something like this, you need to clean up when the react component gets unmounted
```
useEffect(() => {
   const listenere = () => { ... }
   document.addEventListener("keydown", listener)
   return () => document.removeEventListener("keydown", listener)
}, [])
```

File: website/src/components/TutorialComponents/CodeBlock.tsx (hunk 2 above)
Comment (pelikhan) [reply website/src/components/TutorialComponents/CodeBlock.tsx]: move all inlined functions to name handlers in the component body

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 3]
```diff
@@ -106,8 +107,9 @@ function CodeEditor(props: {
         setCode(_code.slice(0, -1));
     }, []);
 
-    useEditable(editorRef, onEditableChange, {
-        disabled: disabled,
+
+    const editObj = useEditable(editorRef, onEditableChange, {
+        disabled: props.disabled || !hasFocus,
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:112]: destruture props in const local variables

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 4]
```diff
@@ -121,19 +123,62 @@ function CodeEditor(props: {
...
+                editor.blur();
+            }
+        }
+    }
+
+
+
+    useEffect(() => {
+        document.addEventListener('keydown',
+            (e) => handleKeydown(e),
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:172]: you are creating a new closure with this function call; so this defeats removeEventListener as it won't be the same closure instance you are trying to remove

**Review by pelikhan — CHANGES_REQUESTED** [review]

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 5]
```diff
@@ -120,20 +125,63 @@ function CodeEditor(props: {
...
+        document.addEventListener('keydown',
+            handleKeydown,
+            {
+                capture: true,
+            }
+        );
+        return () => {
+            document.removeEventListener('keydown', handleKeydown);
+        };
+    }, []);
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:183]: handleKeyDown will depend on "disabled",

- wrap handleKeyDown using useCallback and put `disabled` in the dependency list
- put handleKeyDown in the dependency list of useEffect

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 6]
```diff
@@ -165,20 +213,19 @@ function CodeEditor(props: {
...
-                                // margin: 0,
-                                // outline: "none",
                                 padding: "0",
-                                // fontFamily: "inherit",
-                                // fontSize: "inherit",
                                 ...(!props.className || !props.theme ? {} : _style),
                             }}
                             ref={editorRef}
                             spellCheck="false"
+                            onFocus={(e) => handleFocus(e)}
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:227]: onFocus={handleFocus}

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 7]
```diff
@@ -165,20 +213,19 @@ function CodeEditor(props: {
...
+                            onBlur={(e) => handleBlur(e)}
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:228]: onBlur={handleBlur}

File: website/src/components/TutorialComponents/CodeBlock.tsx (hunk 5 above)
Comment (rlisahuang) [reply website/src/components/TutorialComponents/CodeBlock.tsx]: https://github.com/microsoft/z3guide/pull/109/commits/42853c2a32ed5c7a9649958b661224a3eccfb1fd this?

**Comment by rlisahuang** [comment]
@pelikhan Can I merge this in? It also fixes a rather critical bug (non-playground editors no longer editable)

**Review by pelikhan — APPROVED** [review]

**Comment by pelikhan** [comment]
Always good to add a description to a PR once it is ready for review. (Future you will thank present you)

**Pull request merged**